feat(trade): use Trading SDK in trade flows

shoom3301 opened the pull request:
# Summary

Finally, dog-fooded TradingSDK in CoW Swap.
It required some changes in the SDK: https://github.com/cowprotocol/cow-sdk/pull/256

There are tow big changes:
1. Quote fetching
2. Order posting

Now we have a instance of `TradingSdk` in `src/tradingSdk/tradingSdk` plus `TradingSdkUpdater` for updating appCode, chainId, signer and env in `TradingSdk`.

## Quote fetching
The most of the changes are in this commit: https://github.com/cowprotocol/cowswap/pull/5549/commits/e02652f3a4dfc90455d95014bb67198765f0cd7c
There are other fixes on top of this commit, see them in the final changes.

## Order posting

**That part needs refactoring in the future!**
As we discussed, it requires significant time to make the design better, so I only migrated existing code to TradingSDK.
The plan for the refactoring will be bellow.

You can find key changes by searching `postSwapOrderFromQuote(` and `postLimitOrder(` in the code.
Those calls are located in 6 flow functions:
 - swapFlow
 - ethFlow
 - safeBundleApprovalFlow
 - safeBundleEthFlow
 - limit orders tradeFlow
 - limit orders safeBundleFlow

In the next iterations would be great to merge the first 4 of flows into one.

Basically, I replaced `signEthFlowOrderStep`, `signAndPostOrder`, `buildPresignTx` functions with the calls above.

## Refactoring plan

1. Merge `swapFlow`, `ethFlow`, `safeBundleApprovalFlow`, `safeBundleEthFlow` into one flow in order to reduce code duplications
2. The same for limit order flows
3. Replace `TradeFlowContext` with `tradeQuotesAtom` state. Trade quote already contains all the necessary data and duplicates `TradeFlowContext`
4. Replace `getQuote` from `src/api/cowProtocol` with `TradingSDK`. It's still in use in `FullAmountQuoteUpdater` and `UnfillableOrdersUpdater`

# To Test

All the trading flows:
 - swap (EOA, Safe, SC)
 - limit orders (EOA, Safe, SC)
 - twap (EOA, Safe, SC)
 - eth-flow

Trade parameters:
 - slippage (default, manual, smart)
 - partner fee (cowswap, safe, widget)

Confirmation screen:
 - price updated

Quote:
 - all the parameters (partiallyFillable, receiver, etc.)
 - fast/optimal
 - quote API errors. I'm not sure they are processed well

App data:
 - all the parameters (partner fee, slippage, utm, referrer)

Approvals:
 - regular approvals
 - permits


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new Trading SDK integration, enabling enhanced order posting and quote management.
  - Added a Trading SDK updater component for seamless environment and wallet synchronization.

- **Refactor**
  - Streamlined trade quote structure and management, resulting in more consistent and robust quoting and order flows.
  - Updated order posting, signing, and presigning logic across swap, limit, and bundle flows to use the new SDK.
  - Simplified context and state management for trade and quote data.
  - Removed legacy trade and profile data fetching functions and unused order signing utilities.

- **Bug Fixes**
  - Improved error handling for failed fetches and operator errors, reducing unnecessary retries and enhancing user feedback.

- **Chores**
  - Upgraded core dependencies, including a major update to the Cow Protocol SDK.
  - Removed unused dependencies and legacy code for a cleaner, more maintainable codebase.

- **Tests**
  - Updated and added tests to reflect the new quote and order flow logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shoom3301 commented on package.json:
```diff
@@ -75,14 +75,9 @@
-    "@cowprotocol/cow-sdk": "^5.10.0-RC.0",
+    "@cowprotocol/cow-sdk": "^5.10.2-RC.0",
     "@cowprotocol/ethflowcontract": "cowprotocol/ethflowcontract.git#main-artifacts",
-    "@davatar/react": "1.8.1",
```
Removed unused deps

elena-zh reviewed (changes requested):
Hey @shoom3301 , noice :)

Issues: 
1. No error when sell amount is too small
![smal limit](https://github.com/user-attachments/assets/965a23b9-405c-4050-85a9-9853a22710e0)
![small](https://github.com/user-attachments/assets/38603910-74fb-4106-856b-312a854305bf)
2. 'Insufficient liquidity' should be displayed when a token is not supported
![unsupp token](https://github.com/user-attachments/assets/38292736-70b5-4419-96cd-0a3dda969781)
3. Weird amount is displayed on the progress bar modal after order placement. Weird because it does not correspond to any of these: Sell/Buy amount, Receive including costs, Minimum or maximum sent
![random](https://github.com/user-attachments/assets/f2624167-0190-4cc5-a5d9-f0318cf4d36a)
4. I did not get why my Buy order was multilpied by 10 in the Sell field. 
Steps: 
- connect to a wallet on sepolia
- specify a buy order of buying 10 USDC per COW token
- place the order
- right after fill in 10 into the Buy order again ---> COW token's value in the sell field will be multiplied by 10
![why x10](https://github.com/user-attachments/assets/1aab2664-a5ac-4d76-ab3c-0e517a607c8a)

5. It would be great to stop quoting when a user is offline. Reproducible on Swap, Hooks and TWAP forms. 
![stop quoting when offline](https://github.com/user-attachments/assets/dea43d72-541f-4efd-9124-beaaa4b3980a)
6. Orders are not sent to a recipient when a user fills in it (all order types - swap, limit, twap)
![no recip](https://github.com/user-attachments/assets/7daadda6-bc69-4ebf-82be-058dff359d88)
![recip](https://github.com/user-attachments/assets/937212e5-2a7b-4555-8a17-aeef92bd3672)

6. Order deadline is not updated when change it in the settings for a swap order (ETH or a market one)
7. Limit order: type is Fill or kill instead of PF 
![image](https://github.com/user-attachments/assets/3b92d8a2-4ced-4065-9ff9-7de51236f10f)

8. Permit flow is not working
![permit flow](https://github.com/user-attachments/assets/689112f7-27f6-4ed8-b7ac-4e64dbebf08e)

9. (might be related to the issue above): I've placed an order with a permittable token on Barn. While it was pending, I returned back to the PR preview where the same token pair was in the sell/buy fields, and I've got this crazy behavior. See the [video](https://www.loom.com/share/4732d84a9cb649f786425bb04467fed1)
10. The app goes mad when:
 - connect to a SC walelt (safe/adgent/any)
 - pick a native token into the sell field
![image](https://github.com/user-attachments/assets/7b6ccaa2-ad85-460c-84ce-a662e43feb92)
11. Safe+WC, 2 signers: the app says that the order is cancelled when it is presigned
![presigned](https://github.com/user-attachments/assets/9c2a9a8e-078b-49ea-a0d0-814e914b5aef)
13. 'Insufficient liquidity' error should be displayed when there is no liquidity for a token pair.
12. I was not able to reproduce the case with getting a referrer into the appdata. Could you please provide instructions how to do this?

shoom3301 commented on apps/cowswap-frontend/src/modules/limitOrders/services/tradeFlow/index.ts:
```diff
@@ -35,17 +37,21 @@ export async function tradeFlow(
   const marketLabel = [sellToken.symbol, buyToken.symbol].join(',')
 
+  if (!quoteState.quote) {
+    throw new Error('Quote is undefined in tradeFlow!')
```
In practice it should never be shown.
Anyway, I changed the error text, thanks!

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -1,58 +1,54 @@
-    currentParams.validFor === nextParams.validFor &&
-    currentParams.fromDecimals === nextParams.fromDecimals &&
-    currentParams.toDecimals === nextParams.toDecimals &&
-    currentParams.isEthFlow === nextParams.isEthFlow &&
-    currentParams.chainId === nextParams.chainId
```
There are no such props in the params, they are inside of the SDK now

shoom3301 commented:
@elena-zh Thanks!

1,2,3,4 - fixed
5 - I wasn't able to reproduce that, let's make a call?
6,7 - fixed
8 - can't reproduce
9 - can't reproduce
10 - can't reproduce
11 - can't reproduce
12 - fixed
13 - we actually don't use this field in app-data, please ignore

elena-zh commented:
Hey @shoom3301 , thanks for fixing issues!
Most of them are addressed and also no longer reproducible for me. But still: 
1. 5th case with Off line mode is still reproducible for me. 
steps: 
- Specify an order
- turn off the internet connection
 - change an amount in the field/switch to another network --> you will get tons of quote requests, especially on GC. 
 I have to admit, that partially the same behavior is reproducible on Barn, but, IMO, it would be great to fix. 

2. 11th case with SC wallets is also reproducible to me: 
Steps
- Open the app inside Safe OR connect to is using WC. Connecting to Argent will also work
- OPen Swap page
- Pick ETH (XDAI) into the sell field

3. (new) No 'Price updated' warning on the Confirm swap modal (it was present in the 1st iteration)
4. (new) 'Expired before creation' error state should look better. See the comparison of a PR error message against Barn
![barn](https://github.com/user-attachments/assets/d625f215-82a2-4abc-8f9f-a63e5bec224a)
![exoired before signing](https://github.com/user-attachments/assets/656a1651-e18a-484e-8eda-2fc02729ac9a)
To reproduce:
 - specify an order (2 min duration with 0 slippage)
 - call signing modal in the wallet, but do not sign it for about a minute
 - sigh the order in 30 seconds before its expiration
5. (new) same with 'Insufficient balance' error 
![barn funds](https://github.com/user-attachments/assets/7d7b03f1-8293-4de0-a46a-a370adc1e24b)
![insuff balance](https://github.com/user-attachments/assets/4ee3e525-8dbb-4d8e-90f3-8a8449376d8b)
To reproduce:
 - specify an order with selling WETH
 - call signing modal in the wallet, but do not sign it
 - in another tab, unwrap all weth from your wallet
 - sigh the order
6. Same with 'insufficient allowance' error
- specify an order with selling WETH
 - call signing modal in the wallet, but do not sign it
 - in another tab, revoke WETH approval
 - sigh the order
7. (I did not test it, but I will) Could you please double check the error text when reaching limitation of 50 limit orders creation? 
8. [the issue is on Barn as well: we missed it during the previous refactoring] No error is displayed (at all!) when a user tried to place 2 identical limit orders. To reproduce: 
 - go to Limit orders page
 - slecify sell/buy amounts, a price
 - pick time: custom option (e.g. absolute time in hours/minutes)
 - place 2 orders one after another: only 1 order will be placed, the 2nd one will not be created. But the UI does not show any message
![2 same orders](https://github.com/user-attachments/assets/66ec89ab-754f-4468-bed7-b44b62149939)

9. . [the issue is on Barn as well: we missed it during the previous refactoring] no handling of 'order already executed' state when a user tried to cancel an order that is already filled. To reproduce:
 - place an order
 - while it is opened, send cancel order request to the wallet, but not execute it
 - when the order is filled, press on the sign the cancellation request in the wallet
10. [the issue is on Barn as well: we missed it during the previous refactoring] no handling of 'order already expired' state when a user tried to cancel an order that is already expired. To reproduce:
 - place an order (2 min duration with 0 slippage)
 - while it is opened, send cancel order request to the wallet, but not execute it
 - wait till the order's expiration time comes 
 - press on the sign the cancellation request in the wallet
![exoired before signing](https://github.com/user-attachments/assets/8fadb234-eacd-4211-970f-59f995659e54)

Error messages related to 6-10 cases can be found here 
![errors](https://github.com/user-attachments/assets/6fee21fb-4c85-4658-a384-dde6f15116fc)


Thanks

shoom3301 commented:
@elena-zh thanks a lot!

1. Fixed
2. Fixed
3. Tbh, I didn't meet that case. I saw quote requests, but due to market conditions I didn't get a quote with different amount. Please let me know if it still a case
4. Fixed
5. Fixed
6. Fixed
7. Fixed (See video)
8. Fixed
9. Fixed
10. Fixed

https://github.com/user-attachments/assets/94813d30-5f49-46ec-b73b-b1cabae132bb

elena-zh commented:
Hey @shoom3301 , great, thanks!
However, still there are some issues:
1. case 3: see the [video](https://www.loom.com/share/cf356e2dfa874d5fb58f41114f047c04). Please start on 1 min 50 sec
2. ETH-flow with Safe (inside and WC): the app placed a ETH-order:
 - I specidied a buy order with ETH
 - the app says about bundling (wrap+place order)
 - however, eth-flow TX is sent to the wallet to sign and execute
![no bundle](https://github.com/user-attachments/assets/e4d674f6-2e11-49f2-8e89-f42224e33254)
- when signed, it is [executed](https://dev.explorer.cow.fi/sepolia/orders/0xfc232ea1ca3532170be8bb353834c463ceaf9ded3e2d98b50ea225cf7b34ba9104501b9b1d52e67f6862d157e00d13419d2d6e95ffffffff). it-s type - 'sell' (however, I placed a buy order)
- right after signing the order, another bundle TX request appears, but it fails when I try to sign it
![2nd eth](https://github.com/user-attachments/assets/2ebd9c55-c7b9-4274-a6dc-5c57a52825b0)
3. 9 case from this https://github.com/cowprotocol/cowswap/pull/5549#issuecomment-2766727672 is not fixed
4. 10 case from this https://github.com/cowprotocol/cowswap/pull/5549#issuecomment-2766727672 is not fixed
5. it would be nice to update text when place 2 identical orders to 
![already exists](https://github.com/user-attachments/assets/34943328-d259-46de-a70b-b42df2770e6a)

Thanks

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeFlow/services/safeBundleFlow/safeBundleEthFlow.ts:
```diff
@@ -108,7 +149,7 @@ export async function safeBundleEthFlow(
 
     logTradeFlow(LOG_PREFIX, 'STEP 5: build presign tx')
-    const presignTx = await buildPresignTx({ settlementContract, orderId })
+    const presignTx = await tradingSdk.getPreSignTransaction({ orderId, account })
```
Account as always truthy here, just check types

shoom3301 commented:
@elena-zh thanks!

1. Fixed
2. Fixed (It might hang in "Confirm in your wallet..." state for awhile, but this is a Safe issue, you can observe the same on prod)
3. Please move it to another issue (out of scope)
4. Please move it to another issue (out of scope)
5. Fixed. Just fyi, it seems like backend renamed the error from `DuplicateOrder` to `DuplicatedOrder`. In this case on prod you will see no error

elena-zh reviewed (approved):
Looks good! 

Thank you!

alfetopito commented on apps/cowswap-frontend/src/api/cowProtocol/errors/OperatorError.ts:
```diff
@@ -50,12 +50,14 @@ export enum ApiErrorCodes {
   AppDataHashMismatch = 'AppDataHashMismatch',
   AppdataFromMismatch = 'AppdataFromMismatch',
-  DuplicateOrder = 'There was another identical order already submitted. Please try again.',
+  DuplicateOrder = DuplicateOrderError,
+  DuplicatedOrder = DuplicateOrderError,
```
This should have been fixed already https://github.com/cowprotocol/services/issues/1928

```suggestion
  DuplicatedOrder = 'There was another identical order already submitted. Please try again.',
```

alfetopito commented on apps/cowswap-frontend/src/modules/ethFlow/services/ethFlow/index.ts:
```diff
@@ -81,21 +78,56 @@ export async function ethFlow({
+      orderToSign: unsignedOrder,
+    } = await wrapErrorInOperatorError(() =>
+      tradeQuote
+        .quote!.postSwapOrderFromQuote({
```
I'm not found of non-null assertions, but I don't want to block this change.
Please do come back to it later and throw a explicit error instead.

shoom3301 replied:
Thanks!
Fixed that: https://github.com/cowprotocol/cowswap/pull/5549/commits/e901c98458819c074859ebef7bdfbaffe4ce27e8

alfetopito commented on apps/cowswap-frontend/src/modules/tradeFlow/services/safeBundleFlow/safeBundleApprovalFlow.ts:
```diff
@@ -59,8 +67,35 @@ export async function safeBundleApprovalFlow(
+      orderToSign: unsignedOrder,
+    } = await wrapErrorInOperatorError(() =>
+      tradeQuote
+        .quote!.postSwapOrderFromQuote({
```
With the check above on line 38, is the non-null assertion still needed?

shoom3301 replied:
Thanks!
Fixed that: https://github.com/cowprotocol/cowswap/pull/5549/commits/e901c98458819c074859ebef7bdfbaffe4ce27e8

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeFlow/services/swapFlow/index.ts:
```diff
@@ -73,13 +75,51 @@ export async function swapFlow(
+
+    let presignTxHash: string | null = null
+
+    if (!orderParams.allowsOffchainSigning) {
+      logTradeFlow('SWAP FLOW', 'STEP 5: presign order (optional)')
+      const presignTx = await tradingSdk.getPreSignTransaction({ orderId, account })
+
+      presignTxHash = (await orderParams.signer.sendTransaction(presignTx)).hash
+    }
```
Good point!
I expect it to work the same.
Anyway, at least for unufucation purposes I did a small refactoring:
https://github.com/cowprotocol/cowswap/pull/5549/commits/508f070177df435c8bc565cc7046da33df110527

shoom3301 replied:
@elena-zh could you test Swap and Limit orders with SC wallets (not Safe) please?
I cannot do it right now because the only SC wallet I know is Argent and it went to recovering process for the next 24 hours...

anxolin commented on apps/cowswap-frontend/src/cowSdk.ts:
```diff
@@ -10,4 +12,15 @@ export const metadataApiSDK = new MetadataApi()
+    ...DEFAULT_BACKOFF_OPTIONS,
+    retry(e: any, attemptNumber: number): boolean | Promise<boolean> {
+      // Retry only once when browser is offline
+      if (e?.message === FAILED_FETCH_ERROR && attemptNumber > 1) {
```
can we infer that is offline becasue of failed fetch message?

We also have a subscription to the online state changes, and in principle we stop doing requests.

shoom3301 replied:
Yes, it's actually not relevant anymore, removed it. Thanks!

alfetopito reviewed (approved):
Working well both via EOA and Safe, at least the SWAP and Bundle SWAP flows respectively.
